[docs] basic doc on sponsored transactions (#11090)
aptos-labs/aptos-core: developer-docs-site/docs/guides/sponsored-transactions.md

# Sponsored Transactions

> As outlined in [AIP-39](https://github.com/aptos-foundation/AIPs/blob/main/aips/aip-39.md), a sponsored transactions allow a blockchain account to pay the fees associated with executing a transaction for another account. Sponsored transactions simplify the process for onboarding users into applications by allowing the application to cover all associated fees for interacting with the Aptos blockchain. Here are two examples:

Reviewer: ```suggestion
As outlined in [AIP-39](https://github.com/aptos-foundation/AIPs/blob/main/aips/aip-39.md), sponsored transactions allow an account to pay the fees associated with executing a transaction for another account. Sponsored transactions simplify the process for onboarding users into applications by allowing the application to cover all associated fees for interacting with the Aptos blockchain. Here are two examples:
```

* [MerkleTrade](https://merkle.trade/) offers low cost trading to those with Ethereum wallets by creating an Aptos wallet for users and covering all transaction fees so that the user does not need to acquire utility tokens for Aptos.
* Community engagement applications like [Graffio](https://medium.com/aptoslabs/graffio-web3s-overnight-sensation-81a6cf18b626) offered to cover transaction fees for custodial accounts to support the collaborative drawing application for those without wallets.

## Process Overview

The process for sending a sponsored transaction follows:

> * The sender(s) of the transaction agree upon an operation, as defined by a `RawTransaction`.
> * The sender(s) generate a `RawTransactionWithData::MultiAgentWithFeePayer` structure
>   * As noted below, prior to the framework 1.8 release, this must contain the fee payer's address.
>   * After framework release 1.8, this can optionally be set to `0x0`.
> * The sender(s) can forward the signed transaction to the fee payer who signs with their signature and forwards it to the blockchain.

Reviewer: By definition, there is only one `sender` of a transaction, and multiple `signers` of a transaction.
```suggestion
* The sender of the transaction agrees upon an operation, as defined by a `RawTransaction`.
* The sender generates a `RawTransactionWithData::MultiAgentWithFeePayer`
  * Prior to the framework 1.8 release, this must contain the fee payer's address.
  * After framework release 1.8, this can optionally be set to `0x0`.
* The sender can forward the signed transaction to the fee payer to sign it and forward it to the blockchain.
```

Author: fixed

> * Upon execution of the transaction, the sequence number of the primary sending account is incremented and all gas fees are deducted and all refunds are sent to the gas fee payer.

Reviewer: ```suggestion
* Upon execution of the transaction, the sequence number of the sender account is incremented, all gas fees are deducted from the gas fee payer, and all storage refunds are sent to the gas fee payer.
```

> Alternatively, if the fee payer knows the operation and all parties involved, they could sign a transaction and forward it back to the parties to sign.

Reviewer: Parties is fine here, though I think we should probably stick with common terms, such as signers, rather than move between different terms.

```suggestion
Alternatively, if the fee payer knows the operation and all signers involved, the fee payer could generate and sign the transaction and send it to the other signers to sign.
```

## Technical Details

In Aptos, a sponsored transaction reuses the same SignedTransaction as any other user transaction:
```rust
pub struct SignedTransaction {
    /// The raw transaction
    raw_txn: RawTransaction,

    /// Public key and signature to authenticate
    authenticator: TransactionAuthenticator,
}
```

The difference is in the `TransactionAuthenticator`, which stores the authorization from the fee payer of the transaction to extract utility fees from their account:
```rust
pub enum TransactionAuthenticator {
...
    /// Optional Multi-agent transaction with a fee payer.
    FeePayer {
        sender: AccountAuthenticator,
        secondary_signer_addresses: Vec<AccountAddress>,
        secondary_signers: Vec<AccountAuthenticator>,
        fee_payer_address: AccountAddress,
        fee_payer_signer: AccountAuthenticator,
    },
...
}
```

To prepare a sponsored transaction for an account, the account must first exist on-chain. This is a requirement that is being removed with the 1.8 framework release.

> As of the 1.8 framework release, an account need not exist on-chain. However, the first transaction for an account requires enough gas to not only perform the operation dictated by the transaction but also cover the costs associated with account creation, even if an account already exists. This requirement is expected to subside with advances in our account model.

Reviewer: It's already out for voting, maybe just want to shortcut straight to post 1.8 release?  (if it does get approved)

> As of the 1.8 framework release, an account need not exist on-chain. However, the first transaction for an account requires enough gas to not only perform the operation dictated by the transaction but also cover the costs associated with account creation, even if an account already exists. This requirement is expected to subside with advances in our account model.

Reviewer: Overly wordy / possibly poor to translate

```suggestion
As of the 1.8 framework release, an account does not need to exist on-chain. However, the first transaction for an account requires enough gas to not only perform the transaction and cover costs associated with account creation, even if an account already exists. Future improvements to the account model may remove this requirement.
```

> During signing of the transaction, all parties actually sign the following:

Reviewer: ```suggestion
During signing of the transaction, all parties sign the following:
```

```rust
pub enum RawTransactionWithData {
...
    MultiAgentWithFeePayer {
        raw_txn: RawTransaction,
        secondary_signer_addresses: Vec<AccountAddress>,
        fee_payer_address: AccountAddress,
    },
}
```

> Prior to framework release 1.8, all senders were required to know the actual fee payer address prior to signing. As of framework release 1.8, this restriction has been eliminated and senders can optionally set the address to `0x0` and only the fee payer must sign on their address.

Reviewer: ```suggestion
Prior to framework release 1.8, all signers were required to know the fee payer address prior to signing. As of framework release 1.8, signers can optionally set the address to `0x0` and only the fee payer must sign with their address.
```
